test/functional/sqlsrv/sqlsrv_ae_type_conversion_select.phpt

```diff
@@ -0,0 +1,302 @@
+                            array('x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','y','x','x','x'),//datetimeoffset
+                            array('x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','y','x','x'),//datetime2
+                            array('x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','y','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','y','x'),//decimal
+                            array('x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','x','y'),//numeric
+                            );
+
+set_time_limit(0);
+sqlsrv_configure('WarningsReturnAsErrors', 1);
+
+$connectionInfo = array("CharacterSet"=>"UTF-8");
+$conn = AE\connect($connectionInfo);
+if (!$conn) {
+    fatalError("Could not connect.\n");
+}
+
+$tableName = "type_conversion_table";
+$columns = array();
+$insertQuery = "";
+
+formulateSetupQuery($tableName, $dataTypes, $columns, $insertQuery);
+
+$stmt = AE\createTable($conn, $tableName, $columns);
+if (!$stmt) {
+    fatalError("Failed to create table $tableName\n");
+}
+
+// The data we test against is in values.php
+for ($v = 0; $v < sizeof($values); ++$v)
+{
+    // Each value must be inserted twice because the AE and non-AE column are side by side.
+    $testValues = array();
+    for ($i = 0; $i < sizeof($values[$v]); ++$i) {
+        $testValues[] = $values[$v][$i];
+        $testValues[] = $values[$v][$i];
+    }
+
+    // Insert the data using sqlsrv_prepare()
+    $stmt = sqlsrv_prepare($conn, $insertQuery, $testValues);
+    if ($stmt == false) {
+        print_r(sqlsrv_errors());
+        fatalError("sqlsrv_prepare failed\n");
+    }
+
+    if (!sqlsrv_execute($stmt)) {
+        print_r(sqlsrv_errors());
+        fatalError("sqlsrv_execute failed\n");
+    }
+    
+    sqlsrv_free_stmt($stmt);
+
+    // Formulate the matrix of SELECT queries and iterate over each index.
+    $selectQuery = array();
+    $selectQueryAE = array();
+    formulateSelectQuery($tableName, $selectQuery, $selectQueryAE, $dataTypes);
+
+    for ($i = 0; $i < sizeof($dataTypes); ++$i) {
+        for ($j = 0; $j < sizeof($dataTypes); ++$j) {
+            $stmt = sqlsrv_query($conn, $selectQuery[$i][$j]);
+
+            if ($stmt == false) {
+                $convError = sqlsrv_errors();
+                
+                checkErrors($convError);
+                
+                if (AE\isDataEncrypted()) {
+                    $stmtAE = sqlsrv_query($conn, $selectQueryAE[$i][$j]);
+                    $convError = sqlsrv_errors();
+                    
+                    // if the non-AE conversion fails, certainly the AE conversion
+                    // should fail but only with error 22018.
```

Comment:
Please extend the comment to include the meaning of 22018 - Conversion not allowed - such that we don't need to scroll all the way up to know what it means